File: KotlinLibrary/library/src/main/java/com/lee/library/widget/nav/DotView.java
package com.lee.library.widget.nav;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PaintFlagsDrawFilter;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;

import androidx.core.content.ContextCompat;

import com.lee.library.R;

/**
 * @author jv.lee
 * @date 2019/5/7
 */
public class DotView extends View {
    private Paint mBackgroundPaint;
    private int mWidth;
    private int mHeight;
    private int backgroundColor = Color.RED;


    public DotView(Context context) {
        this(context, null);
    }

    public DotView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public DotView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initPaint();
    }

    private void initPaint() {
        mBackgroundPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mBackgroundPaint.setColor(backgroundColor);
        mBackgroundPaint.setStrokeWidth(1);
        mBackgroundPaint.setStyle(Paint.Style.FILL);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        mWidth = getMeasuredWidth();
        mHeight = getMeasuredHeight();
    }

    @SuppressLint("DrawAllocation")
    @Override
    protected void onDraw(Canvas canvas) {
        canvas.setDrawFilter(new PaintFlagsDrawFilter(0, Paint.ANTI_ALIAS_FLAG));
        drawBackground(canvas);
    }

    private void drawBackground(Canvas canvas) {
        Drawable drawable = ContextCompat.getDrawable(getContext(), R.drawable.shape_indicator_dot);
        if (drawable != null) {
            setBackground(drawable);
        } else {
            mBackgroundPaint.setColor(backgroundColor);
            canvas.drawCircle(mWidth / 2, mHeight / 2, mWidth / 2, mBackgroundPaint);
        }
    }

    @Override
    public void setBackgroundColor(int backgroundColor) {
        this.backgroundColor = backgroundColor;
    }

}

File: KotlinLibraryX/library/src/main/java/com/lee/library/widget/nav/DotView.kt
package com.lee.library.widget.nav

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import androidx.annotation.ColorInt
import com.lee.library.R

/**
 * @author jv.lee
 * @date 2019/5/7
 */
class DotView : View {

    private var mBackgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var mWidth = 0
    private var mHeight = 0
    private var backgroundColor = Color.RED

    constructor(context: Context) : this(context, null, 0)
    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)
    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    ) {
        init(attrs)
    }

    private fun init(attrs: AttributeSet?) {
        attrs?.let {
            context.obtainStyledAttributes(it, R.styleable.DotView).run {
                backgroundColor = getColor(R.styleable.DotView_dotBackground, Color.RED)
                recycle()
            }
        }
        mBackgroundPaint.let {
            it.color = backgroundColor
            it.strokeWidth = 1f
            it.style = Paint.Style.FILL
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        mWidth = measuredWidth
        mHeight = measuredHeight
    }

    override fun onDraw(canvas: Canvas) {
        drawBackground(canvas)
    }

    private fun drawBackground(canvas: Canvas) {
        canvas.drawCircle(
            mWidth / 2.toFloat(),
            mHeight / 2.toFloat(),
            mWidth / 2.toFloat(),
            mBackgroundPaint
        )
    }

    override fun setBackgroundColor(backgroundColor: Int) {
        this.backgroundColor = backgroundColor
    }

    fun setDotColor(@ColorInt color: Int) {
        mBackgroundPaint.color = color
        invalidate()
    }

}